fem/pfespace.cpp

```diff
@@ -1038,263 +1070,635 @@ void ParFiniteElementSpace::ConstructTrueNURBSDofs()
+         ned = fec->DofForGeometry(Geometry::SEGMENT);
+
+         return (index < ghost)
+                ? nvdofs + index*ned + edof // regular edge
+                : ndofs + ngvdofs + (index - ghost)*ned + edof; // ghost edge
+
+      default:
+         ghost = pncmesh->GetNFaces();
+         ned = fec->DofForGeometry(mesh->GetFaceBaseGeometry(0));
+
+         return (index < ghost)
+                ? nvdofs + nedofs + index*ned + edof // regular face
+                : ndofs + ngvdofs + ngedofs + (index - ghost)*ned + edof; //ghost
+   }
+}
+
+/** Dissect a DOF number to obtain the entity type (0=vertex, 1=edge, 2=face),
+ *  entity index and the DOF number within the entity.
+ */
+void ParFiniteElementSpace::UnpackDof(int dof,
+                                      int &entity, int &index, int &edof) const
+{
+   MFEM_VERIFY(dof >= 0, "");
+   if (dof < ndofs)
+   {
+      if (dof < nvdofs) // regular vertex
       {
-         double x = pm(0,i), y = pm(1,i);
-         corner[i] = ((x == 0.0 || x == 1.0) && (y == 0.0 || y == 1.0));
+         int nv = fec->DofForGeometry(Geometry::POINT);
+         entity = 0, index = dof / nv, edof = dof % nv;
+         return;
       }
-      if (nv)
+      dof -= nvdofs;
+      if (dof < nedofs) // regular edge
       {
-         for (i = 0; i < 4; i++)
-         {
-            if (corner[i])
-            {
-               for (k = 0; k < nv; k++) { slave_dofs[i*nv + k] = INVALID_DOF; }
-            }
-         }
+         int ne = fec->DofForGeometry(Geometry::SEGMENT);
+         entity = 1, index = dof / ne, edof = dof % ne;
+         return;
+      }
+      dof -= nedofs;
+      if (dof < nfdofs) // regular face
+      {
+         int nf = fec->DofForGeometry(mesh->GetFaceBaseGeometry(0));
+         entity = 2, index = dof / nf, edof = dof % nf;
+         return;
+      }
+      MFEM_ABORT("Cannot unpack internal DOF");
+   }
+   else
+   {
+      dof -= ndofs;
+      if (dof < ngvdofs) // ghost vertex
+      {
+         int nv = fec->DofForGeometry(Geometry::POINT);
+         entity = 0, index = pncmesh->GetNVertices() + dof / nv, edof = dof % nv;
+         return;
+      }
+      dof -= ngvdofs;
+      if (dof < ngedofs) // ghost edge
+      {
+         int ne = fec->DofForGeometry(Geometry::SEGMENT);
+         entity = 1, index = pncmesh->GetNEdges() + dof / ne, edof = dof % ne;
+         return;
       }
-      if (ne)
+      dof -= ngedofs;
+      if (dof < ngfdofs) // ghost face
       {
-         for (i = 0; i < 4; i++)
+         int nf = fec->DofForGeometry(mesh->GetFaceBaseGeometry(0));
+         entity = 2, index = pncmesh->GetNFaces() + dof / nf, edof = dof % nf;
+         return;
+      }
+      MFEM_ABORT("Out of range DOF.");
+   }
+}
+
+/** Represents an element of the P matrix. The column number is global and
+ *  corresponds to vector dimension 0. The other dimension columns are offset
+ *  by 'stride'.
+ */
+struct PMatrixElement
+{
+   HYPRE_Int column, stride;
```

Comment:
It seems to me that the `stride` can be safely removed - it is the same for all matrix entries. Or, am I missing something?

Reply:
The stride is the same only within one MPI rank. Since the P matrix rows are getting sent to other ranks that have no idea about the 'ndofs' value of the original rank, the 'stride' is necessary to construct the vdim expanded version of the matrix if, ordering == byNODES.

Reply:
Ah, I just realized that the `stride`s (because of the way we order the columns globally) depend on the processor that owns the column.